lib/style.dart

```diff
@@ -481,6 +481,97 @@ enum Display {
   NONE,
 }
 
+abstract class Margin {
+  const Margin();
+
+  double get value => this is InsetMargin ? this.value : 0;
+  }
+
+class AutoMargin extends Margin {
+  const AutoMargin();
+}
+
+class InsetMargin extends Margin {
+  final double value;
+  const InsetMargin(this.value);
+}
+
+class Margins {
+  final Margin? left;
+  final Margin? right;
+  final Margin? top;
+  final Margin? bottom;
+
+  const Margins({ this.left, this.right, this.top, this.bottom });
+
+  /// Auto margins already have a "value" of zero so can be considered collapsed.
+  Margins collapse() => Margins(
+    left: left is AutoMargin ? left : InsetMargin(0),
+    right: right is AutoMargin ? right : InsetMargin(0),
+    top: top is AutoMargin ? top : InsetMargin(0),
+    bottom: bottom is AutoMargin ? bottom : InsetMargin(0),
+  );
+
+  Margins copyWith({ Margin? left, Margin? right, Margin? top, Margin? bottom }) => Margins(
+    left: left ?? this.left,
+    right: right ?? this.right,
+    top: top ?? this.top,
+    bottom: bottom ?? this.bottom,
+  );
+
+  Margins copyWithEdge({ double? left, double? right, double? top, double? bottom  }) => Margins(
+    left: left != null ? InsetMargin(left) : this.left,
+    right: right != null ? InsetMargin(right) : this.right,
+    top: top != null ? InsetMargin(top) : this.top,
+    bottom: bottom != null ? InsetMargin(bottom) : this.bottom,
+  );
+
+  bool get isAutoHorizontal => (left is AutoMargin) || (right is AutoMargin);
+
+  Alignment? get alignment {
```

Comment:
I like where we're going with this PR, but I'm not sure how I feel about the margin class handling its own alignment. I'm sure there are edge cases where these alignments don't hold true. 

For now, can we move this alignment calculation outside of the Margins class?

I'm also going to look into how the w3 says renderers should handle auto margins for inspiration.

Thank you!